### PR TITLE
Add autolet

### PR DESCRIPTION
**Pull request by IamIpanda**
All effects are shown in the test.

NOTES: 
+ Assignment children process missed. I can’t construct a legal case.
+ How to handle type?

### CODE REVIEW

File: source/parser.hera [hunk 1]
```diff
@@ -4147,6 +4148,7 @@ Reset
       set(b) {
         if (b) {
           this.autoVar = true
+          this.autoLet = false
```
Comment (STRd6) [inline source/parser.hera:4151]: This isn't necessary here or in the else branch below since `autoLet` is unrelated to the `coffeeCompat` flags.

Comment (edemaine) [reply source/parser.hera:4151]: `coffeeCompat` implies `autoVar`, right?  How do `autoVar` and `autoLet` interact?  Presumably one should disable the other, maybe whichever is specified last.  So you could `"civet coffeeCompat autoLet"` and `autoLet` would win.

File: source/parser.hera [hunk 2]
```diff
@@ -4161,6 +4162,7 @@ Reset
           this.coffeeOf = true
         } else {
           this.autoVar = false
+          this.autoLet = false
```
Comment (STRd6) [inline source/parser.hera:4165]: This one can be removed too.

File: source/parser.hera (hunk 1 above)
Comment (STRd6) [reply source/parser.hera]: To resolve that we should check and process `autoLet` before `autoVar` inside `processProgram`.

**Comment by IamIpanda** [comment]
Setting `autolet` is removed for `coffeeCompat`.

Add a recursion inner AssignmentExpression to resolve NOTE 1.

**Comment by STRd6** [comment]
@IamIpanda Thanks for the PR, looks great!

I think for `autoLet` types should be handled by TypeScript's type inference. In cases where you need to specify a particular type use an explicit declaration.

This looks good enough to merge in. It's ok for features flagged behind `"civet ..."` directives to be somewhat experimental and open to future improvements.

@edemaine any other thoughts?

**Comment by edemaine** [comment]
I agree!  I'd like to see a short description of what `autoLet` does in the README, though, ideally before merging.

I also had one question: Does `autoLet` still prevent shadowing across functions, as in:
```coffee
x = 5
function f()
  x = 10  # is this the global x?
```

I assume so, like `autoVar`.  I think that's fine, but I wanted to be clear which one it does. (And maybe it's worth adding a test to this effect.)

Assuming yes, in the future, I'd like to make a `shadow` flag where `autoVar` and `autoLet` both don't do this, and limit their automatic scoping to the current function, not parent functions, except when overridden by `global`/`nonlocal` (or `outer`?) directives, like Python. This can lead to fewer nonshadowing bugs (which I've run into maaany times), though it's probably less common with TypeScript.

**Pull request merged**